File: datamodel/src/main/kotlin/de/westermann/robots/datamodel/util/Track.kt
package de.westermann.robots.datamodel.util

import kotlin.math.*

/**
 * @author lars
 */
data class Track(
        val x: Double,
        val y: Double
) {

    val radius: Double = sqrt(x * x + y * y)
    val angle: Double = if (y >= 0) {
        acos(x / radius)
    } else {
        2 * PI - acos(x / radius)
    }

    fun toJson() = json {
        value("x") { x }
        value("y") { y }
    }

    fun toMqtt() = "$x,$y"
    fun normalize(): Track = radius.let {
        if (it > 1.0) {
            copy(x = x / it, y = y / it)
        } else {
            this
        }
    }

    fun discretize(stepSize: Double) = Track(
            (x / stepSize).roundToInt() * stepSize,
            (y / stepSize).roundToInt() * stepSize
    )


    val isZero: Boolean
        get() = x == 0.0 && y == 0.0

    fun radius(r: Double): Track = radius.let { oldRadius ->
        (max(r, 0.0) / oldRadius).let {
            copy(x = x * it, y = y * it)
        }
    }

    companion object {
        val DEFAULT = Track(0.0, 0.0)

        fun parse(data: String) = data.split(",").let {
            Track(it[0].toDouble(), it[1].toDouble())
        }

        fun fromJson(json: Json) = Track(
                json["x"]?.toString()?.toDoubleOrNull() ?: 0.0,
                json["y"]?.toString()?.toDoubleOrNull() ?: 0.0
        )
    }
}

File: python/src/util/track.py
from math import sqrt, acos, pi


class Track:

    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y

    def radius(self) -> float:
        return sqrt(self.x ** 2 + self.y ** 2)

    def angle(self) -> float:
        r = self.radius()
        if r == 0:
            return 0
        elif self.y >= 0:
            return acos(self.x / r)
        else:
            return 2 * pi - acos(self.x / r)

    def __str__(self):
        str(self.x) + "," + str(self.y)

    def normalize(self) -> "Track":
        r = self.radius()
        if r > 1.0:
            return Track(self.x / r, self.y / r)
        return self

    def is_zero(self) -> bool:
        return (self.x == 0) and (self.y == 0)

    def copy_with_radius(self, new_radius: float) -> "Track":
        old_radius = self.radius()
        r = max(new_radius, 0) / old_radius
        return Track(self.x * r, self.y * r)

    @staticmethod
    def parse(value: str) -> "Track":
        h = value.split(',')
        return Track(float(h[0]), float(h[1]))
